### PR TITLE
feat: add sendgrid client module (CMC-954)

### PR DESCRIPTION
**Pull request by sabahirfan**
### JIRA link (if applicable) ###
[CMC-954](https://tools.hmcts.net/jira/browse/CMC-954)

### Change description ###
This PR adds the sendgrid client that we will use to send RPA emails.

**Does this PR introduce a breaking change?** (check one with "x")

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by dharmendrak — DISMISSED** [review]

File: build.gradle [hunk 1]
```diff
@@ -10,14 +10,88 @@ plugins {
...
   id "io.freefair.lombok" version "5.3.0"
   id "org.flywaydb.flyway" version "7.1.1"
+  id 'net.ltgt.apt' version '0.21'
 }
 
 group = 'uk.gov.hmcts.reform'
 version = '0.0.1'
 
 allprojects {
+
```
Comment (swalker125) [inline build.gradle:20]: lots of the changes here seem irrelevant to PR

File: sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                )));
+            }
+        } catch (IOException exception) {
+            throw new EmailSendFailedException(exception);
+        }
+    }
+
+    private boolean is2xxSuccessful(Response response) {
+        return response.getStatusCode() / 100 == 2;
+    }
```
Comment (swalker125) [inline sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java:63]: is there not a method under HttpStatus for this?

File: sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    @SuppressWarnings("unused")
```
Comment (swalker125) [inline sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java]: why?

Comment (swalker125) [reply sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java]: if valid a comment could help this more obvious

File: sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java (hunk 2 above)
Comment (sabahirfan) [reply sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java:63]: Do not want to add the dependency for just little this check.

File: build.gradle (hunk 1 above)
Comment (sabahirfan) [reply build.gradle:20]: really 🤔  which ones?

File: sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    @Retryable(value = EmailSendFailedException.class, backoff = @Backoff(delay = 100, maxDelay = 500))
+    public void sendEmail(String from, EmailData emailData) {
+        verifyData(from, emailData);
+        try {
+            Email sender = new Email(from);
+            String subject = emailData.getSubject();
+            Email recipient = new Email(emailData.getTo());
+            Content content = new Content(TEXT_PLAIN_VALUE, getMessage(emailData));
+            Mail mail = new Mail(sender, subject, recipient, content);
+            if (emailData.hasAttachments()) {
```
Comment (pd-dudkiewicz) [inline sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java]: that check could be removed

File: sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+    @Retryable(value = EmailSendFailedException.class, backoff = @Backoff(delay = 100, maxDelay = 500))
+    public void sendEmail(String from, EmailData emailData) {
+        verifyData(from, emailData);
+        try {
+            Email sender = new Email(from);
+            String subject = emailData.getSubject();
+            Email recipient = new Email(emailData.getTo());
+            Content content = new Content(TEXT_PLAIN_VALUE, getMessage(emailData));
+            Mail mail = new Mail(sender, subject, recipient, content);
```
Comment (pd-dudkiewicz) [inline sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java:38]: This could be extracted to some separate method, like `buildMailFromEmailData`. Maybe even make it part of `EmailData` class? Thoughts?

File: sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+    private static Attachments toSendGridAttachments(EmailAttachment attachment) {
+        try {
+            Attachments.Builder builder = new Attachments.Builder(
+                attachment.getFilename(),
+                attachment.getData().getInputStream()
+            );
+            builder.withType(attachment.getContentType());
+            builder.withDisposition("attachment");
+            return builder.build();
```
Comment (pd-dudkiewicz) [inline sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java]: These could possibly be chained.

File: sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java (hunk 4 above)
Comment (sabahirfan) [reply sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java]: done

File: sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java (hunk 3 above)
Comment (sabahirfan) [reply sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java]: added comment

File: sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java (hunk 5 above)
Comment (sabahirfan) [reply sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java:38]: I will prefer to keep it here, as this is small one and used only here.

File: sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java (hunk 6 above)
Comment (sabahirfan) [reply sendgrid-client/src/main/java/uk/gov/hmcts/reform/sendgrid/SendGridClient.java]: done

**Review by pd-dudkiewicz — APPROVED** [review]

**Review by dharmendrak — APPROVED** [review]

**Review by douglasrice — APPROVED** [review]
